lib/src/model/inheriting_container.dart

```diff
@@ -0,0 +1,617 @@
+// Copyright (c) 2019, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/dart/element/type.dart';
+import 'package:dartdoc/src/element_type.dart';
+import 'package:dartdoc/src/model/class.dart';
+import 'package:dartdoc/src/model/comment_referable.dart';
+import 'package:dartdoc/src/model/extension_target.dart';
+import 'package:dartdoc/src/model/model.dart';
+import 'package:dartdoc/src/model_utils.dart' as model_utils;
+import 'package:dartdoc/src/quiver.dart' as quiver;
+import 'package:meta/meta.dart';
+
+/// An mixin to build an [InheritingContainer] capable of being constructed
+/// with a direct call to a [Constructor] in Dart.
+///
+/// Note that [Constructor]s are not considered to be modifiers so a
+/// [hasModifier] override is not necessary for this mixin.
+mixin Constructable on InheritingContainer {
+  Iterable<Constructor> get constructors => element.constructors
+      .map((e) => ModelElement.from(e, library, packageGraph) as Constructor);
+
+  @override
+  bool get hasPublicConstructors => publicConstructors.isNotEmpty;
```

Comment:
I see that `publicConstructors` is not memoized, so this call may be expensive if made repeatedly. What's funny is that `publicConstructorsSorted` _is_ memoized, so using that one instead of `publicConstructors` would be cheaper on successive calls... not sure if there is an AI here; maybe `hasPublicConstructors` is not called repeatedly...

Reply:
hmmm.   all the has* calls tend to be used a fair bit, but generally only once or at most twice per object.  This one is only used once; the calls that are part of `hasModifiers` are used twice.   I think I will just punt on this as if I mess with this now I'll need to do it again with the NNBD pass as the implementation is different.   But there may be some other nearby areas where we could gain from caching the lists in a smarter way.

Reply:
I decided to cache constructors -- that is used in a few places, and will help.